### PR TITLE
[truth] added new extractor

### LINKED ISSUE #3865: Truth Social
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.05.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

US

### Example URLs

- Single video: https://truthsocial.com/@realDonaldTrump/posts/108346966451100254
- Single video: https://truthsocial.com/@RichSementa/posts/108360011206463048
- Single video: https://truthsocial.com/@KarliBonne/108360308038198927
- Single video: https://truthsocial.com/@TheBeauShow/108359711461231741
- Single video: https://truthsocial.com/@steveinman/108360194781589187
- Single video: https://truthsocial.com/@OAN/108360347038764470
- Single video: https://truthsocial.com/@benny/108357875998831094
- Single video: https://truthsocial.com/@crossroads_josh/108354271636352025
- Single video: https://truthsocial.com/@Kanekoa/108358740797291197
- Single video: https://truthsocial.com/@DevinNunes/108357404310489110
- Single video: https://truthsocial.com/@steveinman/108358989218963312
- Single video: https://truthsocial.com/@intheMatrixxx/108360102617518953
- Single video: https://truthsocial.com/@laralogan/108359772245937961
- Playlist: https://truthsocial.com/@steveinman

### Description

A new social network similar to twitter and based on mastodon I believe. Videos use rumble as a backend and they can be downloaded by using the video source link in yt-dlp, but going to the url in a browser shows it as private.

for example, this truth
[https://truthsocial.com/@realDonaldTrump/posts/108346966451100254](https://truthsocial.com/@realDonaldTrump/posts/108346966451100254) 
links to this video (which is marked as private and is not downloadable)
[https://rumble.com/v15p7on-video-upload-for-108346966451100254.html](https://rumble.com/v15p7on-video-upload-for-108346966451100254.html)
but can be downloaded with yt-dlp using the video source url
[https://sp.rmbl.ws/s8/2/N/W/l/l/NWlle.caa.mp4?u=0&b=0](https://sp.rmbl.ws/s8/2/N/W/l/l/NWlle.caa.mp4?u=0&b=0)

### Verbose log

```shell
yt-dlp -vU https://truthsocial.com/@realDonaldTrump/108346966451100254
[debug] Command-line config: ['-vU', 'https://truthsocial.com/@realDonaldTrump/108346966451100254']
[debug] User config "/Users/me/.config/yt-dlp/config": ['--cookies-from-browser', 'firefox', '--cache-dir', '/Users/me/.config/youtube-dl/', '--user-agent', 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:85.0) Gecko/20100101 Firefox/85.0', '--geo-bypass', '-u', 'PRIVATE', '-p', 'PRIVATE', '-i', '--no-mark-watched', '--no-warnings', '--retries', 'infinite', '--fragment-retries', 'infinite', '--download-archive', '/Users/me/.config/youtube-dl/downloads.txt', '-o', '/Volumes/Movies/Youtubes/%(uploader)s/%(playlist_title)s/%(upload_date)s - %(title)s.%(ext)s', '-f', 'bestvideo[ext=mp4][height<=1080]+bestaudio[ext=m4a][abr<=192]/bestvideo[ext=mp4]+bestaudio[ext=m4a]/bestvideo+bestaudio/best', '--sub-format', 'srt', '--convert-subs', 'srt', '--write-subs', '--write-auto-subs', '--embed-subs', '--compat-options', 'no-keep-subs', '--sub-lang', 'en,ja', '--embed-thumbnail', '--parse-metadata', 'channel:%(meta_artist)s', '--parse-metadata', 'webpage_url:%(meta_comment)s', '--parse-metadata', 'upload_date:%(meta_date>%Y:%m:%d)s', '--embed-metadata', '--embed-chapters', '--xattrs', '--audio-format', 'm4a', '--recode-video', 'mp4', '--match-filter', 'duration >=?60 & !is_live']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.05.18 [b14d52355]
[debug] Compatibility options: no-keep-subs
[debug] Python version 3.10.4 (CPython 64bit) - macOS-12.4-arm64-arm-64bit
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: ffmpeg -bsfs
[debug] exe versions: ffmpeg 5.0.1 (setts), ffprobe 5.0.1
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "/Users/me/Library/Application Support/Firefox/Profiles/xy4v0sei.default-release-1574345941647/cookies.sqlite"
[Cookies] Extracted 799 cookies from firefox
[debug] Proxy map: {}
[debug] Loading archive file '/Users/me/.config/youtube-dl/downloads.txt'
Latest version: 2022.05.18, Current version: 2022.05.18
yt-dlp is up to date (2022.05.18)
[debug] [generic] Extracting URL: https://truthsocial.com/@realDonaldTrump/108346966451100254
[generic] 108346966451100254: Requesting header
[generic] 108346966451100254: Downloading webpage
[generic] 108346966451100254: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://truthsocial.com/@realDonaldTrump/108346966451100254
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/yt-dlp/2022.5.18/libexec/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1395, in wrapper
    return func(self, *args, **kwargs)
  File "/opt/homebrew/Cellar/yt-dlp/2022.5.18/libexec/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1465, in __extract_info
    ie_result = ie.extract(url)
  File "/opt/homebrew/Cellar/yt-dlp/2022.5.18/libexec/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 642, in extract
    ie_result = self._real_extract(url)
  File "/opt/homebrew/Cellar/yt-dlp/2022.5.18/libexec/lib/python3.10/site-packages/yt_dlp/extractor/generic.py", line 4031, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://truthsocial.com/@realDonaldTrump/108346966451100254
```

### PR DESCRIPTION
**Pull request by palewire**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

This patch adds support for Truth Social, the social network founded by former U.S. president Donald J. Trump. His posts are of interest to journalists and other open-source intelligence researchers. This addition would make it easier to collect the videos he posts.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

fix #3865

### CODE REVIEW

File: yt_dlp/extractor/truth.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                'like_count': int,
+            },
+       },
+        {
+            'url': 'https://truthsocial.com/@ProjectVeritasAction/posts/108618228543962049',
+            'md5': 'fd47ba68933f9dce27accc52275be9c3',
+            'info_dict': {
+                'id': '108618228543962049',
+                'ext': 'mp4',
+                'title': "RETRACTO #368: Utah NPR Affiliate RETRACTS False Claim Live On Air Following Veritas' Reporting on Curtis Campaign \n“Nothing I ever do will suffice for these people. They are engaged in conspiracy theories. They are doing precisely the thing they project that I do. Which is they don’t believe in facts, they don’t believe in logic, and they don’t believe in rationality.” - James O’Keefe",
```
Comment (HobbyistDev) [inline yt_dlp/extractor/truth.py:39]: use md5 in title

File: yt_dlp/extractor/truth.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        },
+    ]
+    _GEO_COUNTRIES = ['US']  # The site is only available in the US
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        # Get data from API
+        api_url = 'https://truthsocial.com/api/v1/statuses/' + video_id
+        status = self._download_json(api_url, video_id)
```
Comment (HobbyistDev) [inline yt_dlp/extractor/truth.py:58]: inline api url in status

File: yt_dlp/extractor/truth.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        # Pull out metadata
+        title = strip_or_none(clean_html(status.get('content'))) or self._generic_title(url)
+        timestamp = unified_timestamp(status.get('created_at'))
+        account = status.get('account') or {}
+        uploader = strip_or_none(account.get('display_name'))
+        uploader_id = strip_or_none(account.get('username'))
+        uploader_url = ('https://truthsocial.com/@' + uploader_id) if uploader_id else None
+        repost_count = int_or_none(status.get('reblogs_count'))
+        like_count = int_or_none(status.get('favourites_count'))
+        comment_count = int_or_none(status.get('replies_count'))
```
Comment (HobbyistDev) [inline yt_dlp/extractor/truth.py:72]: inline in return

File: yt_dlp/extractor/truth.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    int_or_none,
+    unified_timestamp,
+    strip_or_none
+)
+
+
+class TruthIE(InfoExtractor):
+    """Extract videos from posts on Donald Trump's truthsocial.com."""
+
+    _VALID_URL = r'https://truthsocial\.com/@[^/]+/posts/(?P<id>[0-9]+)'
```
Comment (HobbyistDev) [inline yt_dlp/extractor/truth.py:14]: use `\d` instead

**Comment by HobbyistDev** [comment]
Please add `fix #3865` in OP

**Comment by palewire** [comment]
@HobbyistDev, thanks for the feedback. I believe I have made all the changes you requested.

File: yt_dlp/extractor/truth.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+        # Set the title, handling case where its too long or empty
+        if len(post) > 40:
+            title = post[:35] + "[...]"
+        elif len(post) == 0:
+            title = self._generic_title(url)
+        else:
+            title = post
+        if uploader:
+            title = '%s - %s' % (uploader, title)
```
Comment (HobbyistDev) [inline yt_dlp/extractor/truth.py:80]: I think it's better to keep title as is. We can add uploader at title in output_template
```suggestion
```

File: yt_dlp/extractor/truth.py [hunk 6]
```diff
@@ -61,15 +61,29 @@ def _real_extract(self, url):
         # Pull out video
         url = status['media_attachments'][0]['url']
 
-        # Return the staff
+        # Return the stuff
         account = status.get('account') or {}
+        uploader = strip_or_none(account.get('display_name'))
```
Comment (HobbyistDev) [inline yt_dlp/extractor/truth.py:68]: return `uploader` in return

**Review by HobbyistDev — CHANGES_REQUESTED** [review]

**Comment by HobbyistDev** [comment]
Btw, i am not a maintainer of this repo, i am just a contributor

**Comment by pukkandan** [comment]
> Btw, i am not a maintainer of this repo, i am just a contributor

Your reviews are good and takes some load off of me. Don't hesitate to do more just because you aren't a maintainer 😄

File: yt_dlp/extractor/truth.py (hunk 6 above)
Comment (palewire) [reply yt_dlp/extractor/truth.py]: Done

File: yt_dlp/extractor/truth.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        )
+
+        # Pull out video
+        url = status['media_attachments'][0]['url']
+
+        # Return the stuff
+        account = status.get('account') or {}
+        uploader = strip_or_none(account.get('display_name'))
+        uploader_id = strip_or_none(account.get('username'))
+        post = strip_or_none(clean_html(status.get('content')))
```
[the remark this answers is not in the document]
Comment (palewire) [reply yt_dlp/extractor/truth.py]: Done

**Comment by palewire** [comment]
done

File: yt_dlp/extractor/truth.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
+from .common import InfoExtractor
+
+from ..utils import (
+    clean_html,
+    int_or_none,
+    unified_timestamp,
+    strip_or_none,
+    traverse_obj
+)
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:9]: sort imports. You can use `isort`

File: yt_dlp/extractor/truth.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+
+class TruthIE(InfoExtractor):
+    """Extract videos from posts on Donald Trump's truthsocial.com."""
+
+    _VALID_URL = r'https://truthsocial\.com/@[^/]+/posts/(?P<id>[\d]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:15]:
```suggestion
    _VALID_URL = r'https?://truthsocial\.com/@[^/]+/posts/(?P<id>\d+)'
```

File: yt_dlp/extractor/truth.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    _TESTS = [
+        {
+            'url': 'https://truthsocial.com/@realDonaldTrump/posts/108779000807761862',
+            'md5': '4a5fb1470c192e493d9efd6f19e514d3',
+            'info_dict': {
+                'id': '108779000807761862',
+                'ext': 'qt',
+                'title': '0d8691160c73d663',
+                'description': '',
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:24]: You sure the description is literally an empty string? Lack of description should be denoted by `None`

File: yt_dlp/extractor/truth.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                'uploader': 'Project Veritas Action',
+                'uploader_id': 'ProjectVeritasAction',
+                'uploader_url': 'https://truthsocial.com/@ProjectVeritasAction',
+                'repost_count': int,
+                'comment_count': int,
+                'like_count': int,
+            },
+        },
+    ]
+    _GEO_COUNTRIES = ['US']  # The site is only available in the US
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:54]: Only set this if XFR works for this site

File: yt_dlp/extractor/truth.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+    def _real_extract(self, url):
+        # Get data from API
+        video_id = self._match_id(url)
+        status = self._download_json(
+            'https://truthsocial.com/api/v1/statuses/' + video_id,
+            video_id
+        )
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:62]: read coding conventions
```suggestion
        video_id = self._match_id(url)
        status = self._download_json(
            f'https://truthsocial.com/api/v1/statuses/{video_id}', video_id)
```

File: yt_dlp/extractor/truth.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+        # Pull out video
+        url = status['media_attachments'][0]['url']
+
+        # Return the stuff
+        uploader_id = strip_or_none(traverse_obj(status, ('account', 'username')))
+        post = strip_or_none(clean_html(status.get('content')))
+
+        # Set the title, handling case where its too long or empty
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:71]: These comments are unnecessary

File: yt_dlp/extractor/truth.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    int_or_none,
+    unified_timestamp,
+    strip_or_none,
+    traverse_obj
+)
+
+
+class TruthIE(InfoExtractor):
+    """Extract videos from posts on Donald Trump's truthsocial.com."""
+
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:14]:
```suggestion
```
Set an IE_DESC if you want instead

File: yt_dlp/extractor/truth.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+            title = post[:35] + "[...]"
+        elif len(post) == 0:
+            title = self._generic_title(url)
+        else:
+            title = post
+
+        return {
+            'id': video_id,
+            'url': url,
+            'title': title,
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:82]:
```suggestion
        return {
            'id': video_id,
            'url': url,
            'title': '',
```
Let core code build title. neither description nor generic_title makes sense here.

File: yt_dlp/extractor/truth.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+            'description': post,
+            'timestamp': unified_timestamp(status.get('created_at')),
+            'uploader': strip_or_none(traverse_obj(status, ('account', 'display_name'))),
+            'uploader_id': uploader_id,
+            'uploader_url': ('https://truthsocial.com/@' + uploader_id) if uploader_id else None,
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:87]:
```suggestion
            'uploader_url': format_field(uploader_id, None, 'https://truthsocial.com/@%s'),
```

File: yt_dlp/extractor/truth.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+        elif len(post) == 0:
+            title = self._generic_title(url)
+        else:
+            title = post
+
+        return {
+            'id': video_id,
+            'url': url,
+            'title': title,
+            'description': post,
```
Comment (pukkandan) [inline yt_dlp/extractor/truth.py:83]: If it's only used once, it probably shouldn't be a var
```suggestion
        uploader_id = strip_or_none(traverse_obj(status, ('account', 'username')))
        
        return {
            'id': video_id,
            'url': status['media_attachments'][0]['url'],
            'title': '',
            'description': strip_or_none(clean_html(status.get('content'))),
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by Lesmiscore** [comment]
I'm not sure why didn't you copy the code from [GabIE](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/gab.py) which can be reused by only changing the hostname

File: yt_dlp/extractor/truth.py (hunk 10 above)
Comment (palewire) [reply yt_dlp/extractor/truth.py]: Fixed.

File: yt_dlp/extractor/truth.py (hunk 11 above)
Comment (palewire) [reply yt_dlp/extractor/truth.py:51]: The site is only available in the United States. I need to use a VPN to get it to work from my current location (Sao Paulo, 
Brazil). So I think the answer to your question is, no, the attribute doesn't seem to work. So I'll remove it.

**Comment by palewire** [comment]
@pukkandan, thank you for the feedback. I believe I've made all the changes you requested.

**Comment by palewire** [comment]
Thank you for supporting this patch. It will be used by professional journalists, I promise you.

**Comment by palewire** [comment]
A backport has been porposed in youtube-dl https://github.com/ytdl-org/youtube-dl/pull/31141

**Pull request merged**